java/java.file.launcher/src/org/netbeans/modules/java/file/launcher/Bundle.properties

```diff
@@ -16,3 +16,6 @@
 # under the License.
 
 OpenIDE-Module-Name=Java File Launcher
+GlobalSettingsPanel.vmLabel.text=VM Options:
```

Comment:
This is a solution for another problem with JEP 477 use in NetBeans. The JEP is in preview, however there is no central way to enable preview for all single source Java files.
I was experimenting to store the individual file properties together with the sources in Git, however .nbattrs are not reflected.
It is far behind acceptable usability to let every student every source file first see with errors and then either wait for hint to enable preview (which does only a half of the job by setting the preview partially also for the parent folder, so other sources in the same folder are never prompted with the hint and they fail to run). Or student can manually type --enable-preview into each single Java file properties.

Reply:
> The JEP is in preview, however there is no central way to enable preview for all single source Java files.

good. This would break things very fast. Enabling preview locks the bytecode to a single JDK feature version. NB itself has only one version of javac in it and supports only one version of any preview JEP at any given release + any file can use a different JDK to launch.

Reply:
I'm not sure how intentionally user-added command line option to single source launched files could "break things very fast". Is it a sarcasm?
I'm trying to find a way for my students to use NetBeans to learn Java. What is your goal?

Reply:
> I'm not sure how intentionally user-added command line option to single source launched files could "break things very fast". Is it a sarcasm?

no sarcasm. Only stating how this will cause non-obvious problems. Esp for those who might be told to quickly add a flag to this text field and then forget about it. Or those who are not aware about the impl details of nb-javac.

> I'm trying to find a way for my students to use NetBeans to learn Java. What is your goal?

In this particular PR I mostly try to make sure the single file launcher fits to the rest of NB UX-wise and that it doesn't break other use cases. Basic project maintainer tasks essentially.

Reply:
@asotona you missed the implementation details of nb-javac comment.  A global enable preview might be useful if it validated that the version of nb-javac and the selected / running JDK were in sync for each file open or run???  If things start breaking randomly when people update NetBeans, _even if running on the same JDK_, then that is confusing.  We have to balance everyone's use cases here - please be polite about that!

Reply:
I'm aware that NetBeans is bundling its own nb-javac and I can hardly imagine someone would like to try preview features of the latest JDK using outdated NetBeans (unless we are talking about EA JDKs and JDK javac should take over nb-javac).
Using the latest NetBeans with the latest JDK features (including the preview ones - like JEP 477) seems to me a valid use case and it is not working out of the box now - that is why all these fixes are needed.

Reply:
So the problematic scenario is when someone use the latest NetBeans to try preview features of an outdated JDK?

Reply:
This conversation becomes ridiculous.
I'm trying to fix show stoppers to enable NetBeans with JDK latest features for future generation Java  developers. And you argue agains with incredible scenarios of possible problems of people trying to scratch their yesterdays head with tomorrows hand.

Reply:
Note that the source launcher does not produce any (on-disk) classfiles, at least as far as I know. And `--source` is no longer mandatory for source launcher, so `--enable-preview` is sufficient. I think this significantly reduces possibility of bigger trouble.

And I think simple `--enable-preview` should work also for nb-javac, at least to the degree preview can work with nb-javac.

Overall, I don't think the danger with having some global setting is so big. We have global settings for both Maven and Ant, and if I put something wrong there, I may get weird results.

What may be not-ideal is that one cannot easily see the effective command line (i.e. global + per-file VM options). It might make sense to add a new read-only property to the properties view showing the effective/full command line. Shouldn't be too difficult (I hope), see here:
https://github.com/apache/netbeans/blob/cd0213d9cbd2e1c6c00cbaf11c4fd48d5ff93e61/java/java.source/src/org/netbeans/modules/java/JavaNode.java#L247

Reply:
Yes, however that would require to move the global settings to a module accessible from both (java.source and java.file.launcher).

Reply:
I've changed the global options are persisted in context of Java Platform module
and added read-only propoerty, however compisitions of the command line seems to be confusing, so just mirroring the global options.

Reply:
> We have global settings for both Maven and Ant, and if I put something wrong there, I may get weird results.

Yes of course but that is the reason I asked for the use case. The difference here is that the preview flag influences build and runtime of all project-less programs, while the maven flags for example are for the maven tooling JVM. There isn't really a use case for setting something like this for maven tooling unless for very specific tasks like testing a plugin or something like that.

I like the idea of showing the global args field the file properties. The output could also print the exec line analog how maven support does it - but this doesn't have to be all implemented in the same PR.